common/optimiser/Optimiser.java

```diff
@@ -56,12 +57,16 @@ public Optimiser() {
     }
 
     public synchronized Status optimise(long timeLimitMillis) {
+        if (hasSolver && constraintsChanged()) {
+            constraints.forEach(constraint -> constraint.initialise(solver));
+            status = Status.NOT_SOLVED;
+        }
         if (isOptimal()) return status;
         else if (!hasSolver) initialiseSolver();
         solver.setTimeLimit(timeLimitMillis);
         status = Status.of(solver.solve(parameters));
-        variables.forEach(OptimiserVariable::recordValue);
-        clearInitialisation();
+        if (status != Status.NOT_SOLVED && status != Status.ERROR)
+            variables.forEach(OptimiserVariable::recordValue);
```

Comment:
we can turn this into a method called `recordValues` which will record the variable values and the objective value together